modules/features/sharing/src/debug/kotlin/au/com/shiftyjelly/pocketcasts/sharing/FFmpegMediaService.kt

```diff
@@ -0,0 +1,91 @@
+import com.arthenica.ffmpegkit.FFmpegKit
+import com.arthenica.ffmpegkit.FFmpegSession
+import com.arthenica.ffmpegkit.FFmpegSessionCompleteCallback
+import com.arthenica.ffmpegkit.ReturnCode
+import java.io.File
+import java.io.IOException
+import kotlin.coroutines.resume
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+internal class FFmpegMediaService(
+    private val context: Context,
+) : MediaService {
+    override suspend fun clipAudio(podcast: Podcast, episode: PodcastEpisode, clipRange: Clip.Range): Result<File> = withContext(Dispatchers.IO) {
+        val outputFile = File(context.cacheDir, "${podcast.title} - ${episode.title} - ${clipRange.start.toHhMmSs()}–${clipRange.end.toHhMmSs()}.mp3")
+        if (outputFile.exists()) {
+            return@withContext Result.success(outputFile)
+        }
+
+        val ffmpegFile = File(context.cacheDir, "ffmpeg-audio-clip.mp3")
+        val command = buildString {
+            val audioSource = episode.downloadedFilePath?.let(::File)?.takeIf(File::exists)?.toString() ?: episode.downloadUrl
+            if (audioSource == null) {
+                return@withContext Result.failure(IOException("No stream found for ${episode.uuid} episode"))
+            }
+
+            append("-i $audioSource ") // Audio stream source
+            val coverFile = convertCoverToJpeg(episode).getOrNull() // Convert covers to JPEG because MP3 doesn't support embedding WebP
+            if (coverFile != null) {
+                append("-i $coverFile ")
+            }
+            append("-ss ${clipRange.startInSeconds} ") // Clip start, must be in seconds or HH:MM:SS.xxx
+            append("-to ${clipRange.endInSeconds} ") // Clip end, must be in seconds or HH:MM:SS.xxx
+            append("-q:a 0 ") // Max audio quality
+            append("-map 0:a:0 ") // Include only the first audio stream, videos can have multiple audio streams
+            if (coverFile != null) {
+                append("-map 1:0 ") // Include the cover stream
+                append("-c:1 copy ") // Copy codec for the cover stream
+            }
+            append("-y ") // Overwrite output file if it already exists
+            append("$ffmpegFile") // Output file
+        }
+
+        executeAsyncCommand(command).mapCatching {
+            if (!ffmpegFile.renameTo(outputFile)) {
+                throw IOException("Failed to rename clip file to output file")
+            }
+            outputFile
+        }
+    }
+
+    private suspend fun convertCoverToJpeg(episode: PodcastEpisode): Result<File> {
+        val outputFile = File(context.cacheDir, "ffmpeg-converted-cover.jpeg")
+        val coverPath = episode.imageUrl ?: "${BuildConfig.SERVER_STATIC_URL}/discover/images/webp/960/${episode.podcastUuid}.webp"
+        val command = "-i $coverPath -y $outputFile"
+        return executeAsyncCommand(command).map { outputFile }
```

Comment:
```suggestion
        return executeAsyncCommand(command).map { outputFile }.getOrNull()
```
Your use of map is clever, I don't think I would of thought of that. Is there a reason you don't return just File? from the method?

Reply:
I don't want to get out of `Result`. I wan to have access to exception downstream for logging purposes.

Reply:
I probably was using it somewhere in chain of `.map()` before and refactored the code. I'll change it to `File?`.

Reply:
Fixed in https://github.com/Automattic/pocket-casts-android/pull/2545/commits/9c3fb637e811ac45812febbb5211c4fabe10c0c4